lib/MPIABI/src/MPIABI.jl

```diff
@@ -0,0 +1,79 @@
+module MPIABI
+
+using Preferences
+
+function known_abis()
+    return (:MicrosoftMPI, :MPICH, :OpenMPI, :MPItrampoline)
+end
+
+# const abi = begin
+#     _abi = @load_preference("abi", nothing)
+#     if isnothing(_abi)
+#        _abi = find_abi()
+#        @set_preference!("abi" => _abi)
+#     end
+#     _abi
+# end
+
+const abi = @load_preference("abi", Sys.iswindows() ? :MicrosoftMPI : :MPICH)
+
+function __init__()
+    if Sys.iswindows()
+        if abi != :MicrosoftMPI
+            @error """
+                The MPI ABI is not compatible with the current platform.
+                Please set the MPI ABI to :MicrosoftMPI.
+            """
+        end
+    end
+end
+
+function set_abi(abi)
+    if abi ∉ known_abis()
+        error("""
+            The MPI ABI $abi is not supported.
+            Please set the MPI ABI to one of the following:
+            $(known_abis())
+        """)
+    end
+    @set_preferences!("abi" => abi)
+    @warn "The MPI abi has changed, you will need to restart Julia for the change to take effect" abi
+
+    if VERSION <= v"1.6.5" || VERSION == v"1.7.0"
+        @warn """
+        Due to a bug in Julia (until 1.6.5 and 1.7.1), setting preferences in transitive dependencies
+        is broken (https://github.com/JuliaPackaging/Preferences.jl/issues/24). As a workaround, we 
+        will now add MPIABI as a direct dependency in your current project.
+        """
+
+        uuid = Preferences.get_uuid(@__MODULE__)
+        project_toml, pkg_name = Preferences.find_first_project_with_uuid(uuid)
```

Comment:
Wouldn't `Pkg.add` but much easier? But you probably don't want to require `Pkg` just for a workaround?

Reply:
I am tempted to fix this in Preferences.jl, which is where I stole this code from.

Reply:
If version 1.6.5 was already released I would just make that the compat xD

Reply:
1.6.5 should be very close: https://discourse.julialang.org/t/julia-1-6-5-testing-period/73068